src/test/shell/bazel/bazel_coverage_test.sh

```diff
@@ -66,29 +62,143 @@ int main(void) {
+    esac
+}
+
+# Asserts if the given expected coverage result is included in the given output
+# file.
+#
+# - expected_coverage The expected result that must be included in the output.
+# - output_file       The location of the coverage output file.
+function assert_coverage_result() {
+    local expected_coverage="${1}"
+    local output_file="${2}"
+
+    # Replace newlines with commas to facilitate the assertion.
+    local expected_coverage_no_newlines=$( echo "$expected_coverage" | tr '\n' ',' )
+    local output_file_no_newlines=$( cat "$output_file" | tr '\n' ',' )
+
+    ( echo $output_file_no_newlines | grep  $expected_coverage_no_newlines ) \
+        || fail "Expected coverage result
+<$expected_coverage>
+was not found in actual coverage report:
+<$( cat $output_file )>"
+}
 
-  ending_part=$(sed -n -e '/PASSED/,$p' $TEST_log)
+# Returns the path of the code coverage report that was generated by Bazel by
+# looking at the current $TEST_log. The method fails if TEST_log does not
+# contain any coverage report for a passed test.
+function get_coverage_file() {
+  local ending_part=$(sed -n -e '/PASSED/,$p' $TEST_log)
 
-  coverage_file_path=$(grep -Eo "/[/a-zA-Z0-9\.\_\-]+\.dat$" <<< "$ending_part")
+  local coverage_file_path=$(grep -Eo "/[/a-zA-Z0-9\.\_\-]+\.dat$" <<< "$ending_part")
   [ -e $coverage_file_path ] || fail "Coverage output file does not exist!"
+  echo "$coverage_file_path"
+}
+
+function test_cc_test_coverage_lcov() {
+  local -r lcov_location=$(which lcov)
+  if [[ ! -x ${lcov_location:-/usr/bin/lcov} ]]; then
+    echo "lcov not installed. Skipping test."
+    return
+  fi
+
+  setup_a_cc_lib_and_t_cc_test
+
+  bazel coverage --test_output=all --build_event_text_file=bep.txt //:t \
+      &>$TEST_log || fail "Coverage for //:t failed"
+
+  local coverage_output_file=$( get_coverage_file )
 
-  # Check if a.cc is in the coverage file
-  assert_contains "^SF:.*a.cc$" "$coverage_file_path"
-  # Check if the only branch in a() has correct coverage:
-  assert_contains "^DA:5,1$" "$coverage_file_path"  # true branch should be taken
-  assert_contains "^DA:7,0$" "$coverage_file_path"  # false branch should not be
+  # Check the expected coverage for a.cc in the coverage file.
+  assert_coverage_result "$(get_expected_coverage a.cc)" "$coverage_output_file"
+  # Check the expected coverage for a.cc in the coverage file.
+  assert_coverage_result "$(get_expected_coverage t.cc)" "$coverage_output_file"
 
+  # Verify that this is also true for cached coverage actions.
+  bazel coverage --test_output=all --build_event_text_file=bep.txt //:t \
+      &>$TEST_log || fail "Coverage for //:t failed"
+  expect_log '//:t.*cached'
   # Verify the files are reported correctly in the build event protocol.
   assert_contains 'name: "test.lcov"' bep.txt
   assert_contains 'name: "baseline.lcov"' bep.txt
+}
+
+function test_cc_test_coverage_gcov() {
+  local -r lcov_location=$(which lcov)
```

Comment:
You need lcov for gcov? I think a comment could be helpful here and maybe the check should be part of the setup?